zpay32/invoice_internal_test.go

```diff
@@ -230,13 +296,510 @@ func TestParseTimestamp(t *testing.T) {
+		},
+		{
+			// The first byte is <= 3 bits long.
+			data:   []byte{0x1, 0x2},
+			valid:  true,
+			result: 34,
+		},
+		{
+			// The first byte is > 3 bits long.
+			data:   []byte{0xa, 0x0},
+			valid:  true,
+			result: 320,
+		},
+	}
+
+	for i, test := range tests {
+		length, err := parseFieldDataLength(test.data)
+		if (err == nil) != test.valid {
+			t.Errorf("field data length decoding test %d failed: %v", i, err)
+			return
+		}
+		if test.valid && length != test.result {
+			t.Errorf("%d) failed decoding field data length: "+
+				"expected %d, got %d",
+				i, test.result, length)
+			return
+		}
+	}
+}
+
+// TestParsePaymentHash checks that the payment hash is properly parsed.
+func TestParsePaymentHash(t *testing.T) {
+	t.Parallel()
+
+	var testPaymentHash [32]byte
+	copy(testPaymentHash[:], testPaymentHashSlice[:])
+
+	tests := []struct {
+		data   []byte
+		valid  bool
+		result *[32]byte
+	}{
+		{
+			data:  []byte{},
+			valid: false, // empty data
+		},
+		{
+			data:  []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+			valid: false, // data too short, not 52 bytes
+		},
+		{
+			data:   []byte{0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x4, 0x1, 0x0},
+			valid:  true,
+			result: &testPaymentHash,
+		},
+		{
+			// This test is the same as above, but we append an
+			// extra byte to the data to cause a failure.
+			data:  []byte{0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x0, 0x0, 0x10, 0x4, 0x0, 0x18, 0x4, 0x0, 0x14, 0x3, 0x0, 0xe, 0x2, 0x0, 0x9, 0x0, 0x4, 0x1, 0x0, 0x0},
+			valid: false, // data too long, not 52 bytes
+		},
+	}
+
+	for i, test := range tests {
+		paymentHash, err := parsePaymentHash(test.data)
+		if (err == nil) != test.valid {
+			t.Errorf("payment hash decoding test %d failed: %v", i, err)
 			return
 		}
+		if test.valid && !compareHashes(paymentHash, test.result) {
+			t.Errorf("%d) failed decoding payment hash: "+
+				"expected %x, got %x",
+				i, *test.result, *paymentHash)
+			return
+		}
+	}
+}
+
+// TestParseDescription checks that the description is properly parsed.
+func TestParseDescription(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		data   []byte
+		valid  bool
+		result *string
+	}{
+		{
+			data:   []byte{},
+			valid:  true,
+			result: &testEmptyString,
+		},
+		{
+			data:   []byte{0xa, 0x1, 0x16, 0x6, 0xa, 0x18, 0xb, 0x13, 0xc, 0x14, 0x10, 0x6, 0x6, 0x1b, 0x1b, 0xe, 0xe, 0xd, 0x14, 0x16, 0x8, 0x19, 0xb, 0x12, 0x4, 0x1, 0x19, 0x17, 0xa, 0x1c, 0x3, 0x10, 0xd, 0x1d, 0x19, 0x7, 0x8, 0x1a, 0xb, 0xe, 0xc, 0x1c, 0x10, 0x7, 0x8, 0x1a, 0x3, 0x9, 0xe, 0xc, 0x10, 0x7, 0x0, 0x1c, 0x13, 0xf, 0xd, 0x9, 0x12, 0x16, 0x6, 0x1d, 0x0},
```

Comment:
same, use the expected string -> bytes -> convert to base32. 

Or did you get this data slice from somewhere?